shell/platform/common/cpp/accessibility_bridge.h

```diff
@@ -0,0 +1,240 @@
+class AccessibilityBridge : public AXTreeObserver {
+ public:
+  //------------------------------------------------------------------------------
+  /// Delegate to handle accessibility event and route accessibility action
+  /// back to the Flutter framework.
+  ///
+  /// The accessibility events are generated when accessibility tree changes.
+  /// These events must be sent to the native accessibility system through
+  /// the native API for the system to pick up the changes
+  /// (e.g. NSAccessibilityPostNotification in MacOS).
+  ///
+  /// The accessibility actions are generated by the native accessibility system
+  /// when users interacted with the assistive technologies. Those actions
+  /// needed to be sent to the Flutter framework.
+  class AccessibilityBridgeDelegate {
+   public:
+    virtual ~AccessibilityBridgeDelegate() = default;
+    //------------------------------------------------------------------------------
+    /// @brief      Handle accessibility events generated due to accessibility
+    /// tree
+    ///             changes.
+    ///
+    /// @param[in]  targeted_event      The object that contains both the
+    /// generated
+    ///                                 event and the event target.
+    /// @param[in]  bridge              The pointer to the accessibility bridge
+    /// that
+    ///                                 can be used for querying the
+    ///                                 accessibility information at the time
+    ///                                 the event is fired.
+    virtual void OnAccessibilityEvent(
+        AXEventGenerator::TargetedEvent targeted_event,
+        AccessibilityBridge* bridge) = 0;
+
+    //------------------------------------------------------------------------------
+    /// @brief      Dispatch accessibility action back to the Flutter framework
+    ///
+    /// @param[in]  target              The semantics node id of the action
+    /// target.
+    /// @param[in]  action              The generated flutter semantics action.
+    /// @param[in]  data                Additional data associated with the
+    /// action.
+    /// @param[in]  data_size           The length of the additional data.
+    virtual void DispatchAccessibilityAction(uint16_t target,
+                                             FlutterSemanticsAction action,
+                                             uint8_t* data,
+                                             size_t data_size) = 0;
+  };
+  //------------------------------------------------------------------------------
+  /// @brief      Creates a new instance of a accessibility bridge.
+  ///
+  /// @param[in]  user_data           A custom pointer to the data of your
+  ///                                 choice. This pointer can be retrieve later
+  ///                                 through GetUserData().
+  AccessibilityBridge(std::unique_ptr<AccessibilityBridgeDelegate> delegate,
+                      void* user_data);
```

Comment:
Why is this needed? Can we avoid it?

This feels like something you'd see in a C API rather than C++. It also feels like it's probably introducing some kind of layering violation or incompatibliity. It should probably be some kind of object/class.

Reply:
when embedding implementing the FlutterAccessibility subclass, they may need reference to the native object, for example they may need window rect to properly transforming rect of the accessibility object. In this case, the can only get it from the native FlutterEngine. It is similar for FlutterEngineRun which takes in a user_data, so the embedding can inject their dependency to be used later.